nix-bundle.sh

```diff
@@ -1,56 +1,165 @@
+      --arg <NAME> <EXPR>       argument to be passed to Nix functions
+      --argstr <NAME> <STRING>  string-valued argument to be passed to Nix functions
+  -f, --file <FILE>             evaluate FILE rather than the default
+  -I, --include <PATH>          add a path to the list of locations used to look up <...> file names
+      --bin <BINARY>            the binary to bundle, relative to TARGET's output path; defaults to
+                                the lexicographically first file inside /bin
 
+Examples:
+
+  To bundle GNU Hello from NixOS 19.03:
+  $ $0 -f channel:nixos-19.03 hello --bin /bin/hello
+
+  To bundle the executable from within a project that has a default.nix for build:
+  $ $0 --bin /bin/name-of-the-executable
 EOF
+            exit 0
+            ;;
+        --)
+            shift
+            break
+            ;;
+        -*)
+            echo "error: unrecognized flag '$1'" >&2
+            fail
+            ;;
+        *)
+            targets+=("$1")
+            shift
+            ;;
+    esac
+done
 
-    exit 1
-fi
+while [ $# -ne 0 ]
+do
+    targets+=("$1")
+    shift
+done
 
-nix_file=`dirname $0`/default.nix
+# Create tmp directory for the links produced by `nix build`
 
-target="$1"
-shift
+linkdir=$(mktemp -d)
+build_flags+=("--out-link")
+build_flags+=("$linkdir/result")
 
-extraTargets=
-if [ "$#" -gt 1 ]; then
-    while [ "$#" -gt 1 ]; do
-        extraTargets="$extraTargets $1"
-        shift
-    done
+# Build targets
+
+target_path=
+extra_target_paths=
+
+for target in "${targets[@]}"
+do
+    nix build "${build_flags[@]}" "$target" || exit $?
+    if [ -z "$target_path" ]
+    then
+        target_path="$(readlink "$linkdir/result")"
+    else
+        extra_target_paths="$extra_target_paths $(readlink "$linkdir/result")"
+    fi
+done
+
+if [ -z "$target_path" ]
+then
+    nix build "${build_flags[@]}" || exit $?
+    target_path="$(readlink "$linkdir/result")"
 fi
 
-exec="$1"
-shift
+# Apply default for --bin
+
+if [ -z ${bin+x} ]
+then
+    bin=$(ls "$target_path/bin" | head -1)
+fi
+
+# Determine bootstrap function -- This seems like a total hack!
 
 bootstrap=nix-bootstrap
-if [ "$target" = "nix-bundle" ] || [ "$target" = "nixStable" ] || [ "$target" = "nixUnstable" ] || [ "$target" = "nix" ]; then
+if [ "${targets[0]}" = "nix-bundle" ] || [ "${targets[0]}" = "nixStable" ] || [ "${targets[0]}" = "nixUnstable" ] || [ "${targets[0]}" = "nix" ]; then
     bootstrap=nix-bootstrap-nix
-elif ! [ -z "$extraTargets" ]; then
+elif ! [ -z "$extra_target_paths" ]; then
     bootstrap=nix-bootstrap-path
 fi
 
-expr="with import <nixpkgs> {}; with import $nix_file {}; $bootstrap { target = $target; extraTargets = [ $extraTargets ]; run = \"$exec\"; }"
+# Run bootstrap function
+
+expr="(with import $nix_file {}; $bootstrap { target = $target_path; extraTargets = [$extra_target_paths ]; run = \"$bin\"; })"
+
+echo $expr
 
 out=$(nix-store --no-gc-warning -r $(nix-instantiate --no-gc-warning -E "$expr"))
```

Comment:
This should normally do the actual build, The hope is that it avoids creating ```/result``` dirs in the user's directory.

Reply:
Although it's not clear how to handle some of the new flags like ```-f```, etc. ```nix-instantiate``` does allow ```--arg``` and ```--argstr``` however.

Reply:
`nix build` has nicer output and it allows to use the modern nix cli. The `result` link is created in a temporary directory.